packages/flutter/test/gestures/tap_test.dart

```diff
@@ -398,10 +397,74 @@ void main() {
       'tapA onTapDown',
       'tapA onTapUp',
       'tapA onTap',
-      'tapB onTapCancel',
       'swept 2',
       'disposed A',
       'disposed B',
     ]);
   });
+
+  testGesture('PointerCancelEvent cancels tap', (GestureTester tester) {
+    const PointerDownEvent down = PointerDownEvent(
+        pointer: 5,
+        position: Offset(10.0, 10.0)
+    );
+    const PointerCancelEvent cancel = PointerCancelEvent(
+        pointer: 5,
+        position: Offset(10.0, 10.0)
+    );
+
+    final TapGestureRecognizer tap = TapGestureRecognizer();
+
+    final List<String> recognized = <String>[];
+    tap.onTapDown = (_) {
+      recognized.add('down');
+    };
+    tap.onTapUp = (_) {
+      recognized.add('up');
+    };
+    tap.onTap = () {
+      recognized.add('tap');
+    };
+    tap.onTapCancel = () {
+      recognized.add('cancel');
+    };
+
+    tap.addPointer(down);
+    tester.closeArena(5);
+    tester.async.elapse(const Duration(milliseconds: 5000));
+    expect(recognized, <String>['down']);
+    tester.route(cancel);
+    expect(recognized, <String>['down', 'cancel']);
+
+    tap.dispose();
+  });
+
+  testGesture('losing tap gesture recognizer does not send onTapCancel', (GestureTester tester) {
+    final TapGestureRecognizer tap = TapGestureRecognizer();
+    final HorizontalDragGestureRecognizer drag = HorizontalDragGestureRecognizer();
+
+    final List<String> recognized = <String>[];
+    tap.onTapDown = (_) {
+      recognized.add('down');
+    };
+    tap.onTapUp = (_) {
+      recognized.add('up');
+    };
+    tap.onTap = () {
+      recognized.add('tap');
+    };
+    tap.onTapCancel = () {
+      recognized.add('cancel');
+    };
+
+    tap.addPointer(down3);
```

Comment:
I assume that pointer 3 in this test or pointer 5 in the previous isn't significant; could be any one pointer?

Otherwise: same comment as the previous test. This doesn't look wrong but it's hard to understand if it's right.